### PR TITLE
Use composite action for issue creation

### PR DESCRIPTION
**Pull request by mre**

### CODE REVIEW

**Comment by polarathene** [comment]
You may want to also add [`micalevisk/find-last-issue`](https://github.com/marketplace/actions/find-last-issue) to support the issue feature? (_to handle updating an existing open issue_)

I have [looked into it myself for issue creation with `peter-evans/create-issue-from-file`](https://github.com/peter-evans/create-issue-from-file/issues/298#issuecomment-1173048855) to create or update an existing issue. It works well, see the linked discussion for a snippet example of the two used together.

**Comment by mre** [comment]
> You may want to also add [micalevisk/find-last-issue](https://github.com/marketplace/actions/find-last-issue) to support the issue feature? (to handle updating an existing open issue)

Sorry, not sure I follow. `create-issue-from-file` already supports updating an existing issue. Did you mean that we could use `find-last-issue` to add a new _comment_ to an existing issue? That would be what I had in mind actually. I thought `create-issue-from-file` could do that, but maybe I was wrong.
What I want is for the user to pass an optional issue number an the `lychee-action` would add a comment there. If no issue number was given, it would create a new issue.

**Comment by polarathene** [comment]
> What I want is for the user to pass an optional issue number an the `lychee-action` would add a comment there. If no issue number was given, it would create a new issue.

What is the use of the comment? The author of `create-issue-from-file` has a similar action for managing comments. I use [this action](https://github.com/marocchino/sticky-pull-request-comment) myself for creating and updating a comment on PRs.

- `create-issue-from-file` will create an issue unless given an issue number to update.
- `find-last-issue` provides a means to lookup the issue number of that created issue, instead of updating your workflow (temporarily?) to tell `create-issue-form-file` to update the existing issue and not create a new one.
- This can additionally take the state of an issue into account as well, so that only an open issue is updated... Thus if you resolve the links reported and close the issue, a new one will be opened next time instead until that one is closed. Otherwise the issue will just have it's body content updated/replaced (_no need to append comments?_).

**Comment by mre** [comment]
Gotcha.

If possible I'd like to keep the number of dependencies to other Github actions at a minimum.
This is to avoid additional maintenance overhead when these dependencies change or get deprecated.

I'd love to have the possibility to append a new comment to an existing issue. However this can be added
in a separate PR.
For now I'd like to merge the functionality in as is. Does it look good from your perspective @polarathene?
Would like to get an approval to be sure I didn't mess up.

Also, we should release this as part of version 2.0 of the extension to avoid any breaking changes.
(Otherwise two issues would be created in the worst case &mdash; one by the existing call to `create-issue-from-file` and one by the now built-in issue creator.)

File: action.yml
Comment (polarathene) [inline action.yml:59]: This appears to apply to the entire `env` block? The comment should be above `env:` line then, and perhaps reference the comment too? I assume it was [this one](https://github.com/actions/runner/issues/665#issuecomment-676581170).

File: action.yml
Comment (polarathene) [inline action.yml:60]: What is the point of this ENV? You do not appear to use `INPUT_GITHUB_TOKEN` in your `entrypoint.sh`?

File: action.yml
Comment (polarathene) [inline action.yml:34]: May be better to scope the issue with a prefix that indicates in the title it's a CI issue, and more specific hint of the content (_since while it does contain a report, it's only being reported when failures are caught?_):

```suggestion
    default: 'ci(lint): Some monitored links are failing'
```

File: README.md
Comment (polarathene) [inline README.md:43]:
```suggestion
When failing links are found, `lychee-action` can create and update an issue with the `createIssue` input:

```yaml
- name: Link Checker
  uses: lycheeverse/lychee-action@v2.0.0
  with:
    createIssue: true
```

You might want to show off / mention other supported features too?

**Review by polarathene — CHANGES_REQUESTED** [review]
I have not reviewed `workflows/{links,main}.yml` which appear to be for your own testing of the action.

I realize the review is quite verbose, but hopefully worthwhile :sweat_smile: (_I tried to make it more terse by collapsing less important details_)

**Comment by polarathene** [comment]
> I'd love to have the possibility to append a new comment to an existing issue. However this can be added
in a separate PR.

I'm still curious what the purpose for that would be, vs replacing the automated issue body.

It's easy enough to do regardless, I have a workflow for PRs that adds a comment to the PR with a link to a deployment preview of docs. Whenever the PR is updated, the new deployment preview link is updated in that comment (_the action can alternatively delete the old comment and create a new one_).

---

> If possible I'd like to keep the number of dependencies to other Github actions at a minimum.
> This is to avoid additional maintenance overhead when these dependencies change or get deprecated.

Wouldn't it be better to avoid the composite action entirely then? Or offer that separately from `lychee-action` (_eg: `lychee-action-reporter` as a composited variant?_)

- Is all this worth maintaining vs a little more verbosity for a workflow job you'd rarely need to update (_beyond the version bumps_).
- Will it risk burdening you with more maintenance/support queries vs just documenting an example of a common combination of actions like others do? (_which is unofficial support, that you can more easily delegate to community to contribute/improve_)

---

> For now I'd like to merge the functionality in as is. Does it look good from your perspective @polarathene?
Would like to get an approval to be sure I didn't mess up.

Personally after providing that review for you, my opinion is this is not worth maintaining compared to having better documentation. Is there a strong demand for the composited action?

Provide an `examples/` folder that shares common workflows, and optionally documents them well enough to grok with minimal experience. Users can help contribute to that, and the likelihood is probably more common than contributors to this nested composited approach.

We know of some common examples to cater for already:

- PRs as a lint, fail when link errors are encountered. Optionally demonstrate using another action earlier to provide only inputs being changed in the PR (_a complimentary markdown doc can detail when that isn't ideal, as well as why changed content instead of files can be less reliable too_).
- Scheduled jobs that report failures by creating or updating an existing open issue.
  - Also useful for broader link checking when rate-limit is a concern, as not only could the workflow cache be leveraged (_in addition to `lychee`'s own cache when that is fixed_), but you're more likely to have a higher rate-limit budget vs PRs (_no secrets such as `GITHUB_TOKEN` available for third-party contributor PRs, requires some workarounds to support, otherwise due to 60 requests/hour should exclude github service URLs_).
- API specific workarounds. I'm not sure about this one, but it's good to raise awareness. If someone wants to contribute their own workflow solutions for that, it could be helpful. One example might be for static sites, if deploying to GH Pages, rather than testing against the deployment, checkout the `gh-pages` branch and have `lychee` process the content there instead.

---

> Also, we should release this as part of version 2.0 of the extension to avoid any breaking changes.
> (Otherwise two issues would be created in the worst case — one by the existing call to create-issue-from-file and one by the now built-in issue creator.)

Sure, but I really don't think issue creation should be opt-out.

**Comment by polarathene** [comment]
> Gotcha.

How do you plan for the issue number feature to be used? Currently it seems like:

- Create a new issue each time the action is run and failures are found.
- Or user creates an issue manually, or gets the issue number from the first automated issue creation and updates their workflow to use that.

Both of those seem awkward..? Whereas with `last-issue-action`, you don't have to worry about either of those.

Assuming `lychee-action` completes with an output of `exit_code == 2` (failed links):

- **Create an issue** when one doesn't already exist (_or if one does but since closed_)
- Or **update an issue** that was already created for `lychee-action` (_and still in an open state_).
- **Close the open issue** that was created for `lychee-action` when a PR or similar event closes the issue as resolved (_automatic with `Fixes` or similar syntax in PRs upon merging them_).
- Next time `lychee-action` finds failures, a new issue is opened (_the earlier steps repeat_), and like before it will be updated automatically.
- There is no manual/explicit changes required, nor any issue creation spam (_in the sense of redundantly opening issues each time_). That seems much nicer right?

If you did not visit the link I provided, this is what it'd look like (_ignoring the composited action approach in this PR_):

```yaml
name: 'Manually trigger action (for testing)'
on: workflow_dispatch

jobs:
  lint-links-in-docs:
    name: 'Lint - Documentation Links (with issue reporting)'
    runs-on: ubuntu-22.04
    permissions:
      contents: read
      issues: write
    steps:
      # Permissions (contents: read)
      - uses: actions/checkout@v3

      - name: 'Check links in docs'
        id: lychee
        uses: lycheeverse/lychee-action@v1.5.0
        continue-on-error: true
        with:
          args: --verbose --no-progress -- docs/**/*
          output: ${{ runner.temp }}/lychee-report.md

      # Below manages issue creation/updating when failing links are found:

      # Permissions (issues: read)
      - name: 'Check for an existing lychee-action issue'
        id: last-issue
        uses: micalevisk/last-issue-action@v1.2
        # NOTE: This env won't be necessary soon:
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        # Find the last updated open issue with a `ci-lint-links` label:
        with:
          state: open
          labels: ci-lint-links

      # Permissions (issues: write)
      - name: 'Create a new issue, or update an existing one'
        if: ${{ steps.lychee.outputs.exit_code == 2 }}
        uses: peter-evans/create-issue-from-file@v4
        with:
          title: 'ci(lint): Docs have failing links'
          content-filepath: ${{ runner.temp }}/lychee-report.md
          # If issue_number is empty (no match found), create a new issue, otherwise update this matched issue:
          issue-number: ${{ steps.last-issue.outputs.issue_number }}
          # Add a label unique enough to search for, and any other labels relevant to maintainers:
          labels: ci-lint-links, docs, bug
```

Well.. that would work, except **`lychee-action` doesn't have any outputs available to other steps** because it's only available to the other steps in it's current `composite` action form that I see is already in use prior to this PR :sweat_smile: 

To verify, just swap out those last two steps of the example to view the available outputs. It will show that it's empty even when output `exit_code == 2`. Which means the `if` condition I have in the example will never be true until that problem is resolved.

```yaml
      - name: 'List lychee outputs'
        run: |
          echo '::echo::on'
          echo -e "lychee.outputs:\n${{ toJSON(steps.lychee.outputs) }}"
```

**Comment by polarathene** [comment]
> You are right in that coupling issue creation with this action is probably not the way to go. I've changed my mind.

:+1: 

> Instead, let's work on the examples folder as you suggested.

I can contribute these, and refactor `entrypoint.sh` (_or the action itself due to the `exit_code` output not being available to other steps_).

In exchange, could you please redirect your time on resolving the `lychee --cache` bug?

**Comment by mre** [comment]
Please do, that would be awesome!

> In exchange, could you please redirect your time on resolving the lychee --cache bug?

Sounds good.

**Comment by mre** [comment]
> I can contribute these, and refactor entrypoint.sh (or the action itself due to the exit_code output not being available to other steps).

@polarathene do you still plan to work on the example folder? The exit code should be propagated correctly now.
In any case, I guess we can close this?

**Comment by polarathene** [comment]
> @polarathene do you still plan to work on the example folder?

Yes, sorry been rather busy lately :sweat_smile: 

> The exit code should be propagated correctly now.
> In any case, I guess we can close this?

:+1:

**Comment by tooomm** [comment]
I got lost in the conversation here.
Are there some examples in place that explain how to setup the action to update already existing issues?

**Comment by polarathene** [comment]
> Are there some examples in place that explain

Any particular examples you're after?

I'm unfortunately swamped right now, contributing examples is still on my todo list but I have deadlines to meet :sweat_smile:

**Comment by tooomm** [comment]
>Any particular examples you're after?

I want to automatically create an issue once errors are found with lychee-action.
The issue should be updated once new information is available (add new errors, or remove fixed ones).
The issue should be closed once all errors are resolved.

My first impression was that it should work already, one just needs to know the correct configuration which there are no examples of yet. Hence my question.
But it looks like this does currently not work as lychee-action is not providing any outputs to other steps for now?
Maybe this is better off in an issue as long as it's not working?

>I'm unfortunately swamped right now, contributing examples is still on my todo list but I have deadlines to meet 😅

Great to hear that you still plan on working on it! 👍

**Comment by polarathene** [comment]
> Maybe this is better off in an issue as long as it's not working?

Open an issue and mention me in it, it can track the request for that example.

> My first impression was that it should work already, one just needs to know the correct configuration which there are no examples of yet.

I believe I had something similar to what you're asking for when I was working on a similar workflow. I didn't manage to complete it as it relied on some other actions to update, that should be taken care of now.

I hope you don't mind the delay, but I should be able to find time to sort that out for you this month. If I can get on top of things this week, it might be possible to have a workflow example available this weekend :+1: 

We decided that `lychee-action` should be focused on what it does best, and delegate other parts to actions that do those things well, so you'll have a workflow that uses a few different actions together to achieve this.

---

This is a copy/paste from a workflow I shared elsewhere which should be able to be revised to meet your requirements.

This was while working on a PR for `last-issue` to get some changes for the `2.0` release that I requested, it may need some other changes. `link-reporter` would be `lychee-action` but was stubbed out to test the exit code behaviour on the other two actions for issue management, which you would use to achieve your goal.

If I remember correctly, this would create an issue when `link-reporter` outputs the markdown file with exit code of `2` (failing links), `last-issue` action prior to that would provide an existing open issue to update instead if available.

So this just needs to consider a successful run of `lychee-action` to close any issue that `last-issue` might report as open. Then the next time, a new issue is created and it repeats (_old closed issue is not re-cycled, which I think is the right approach if any discussion is involved_).

```yaml
name: 'Testing - Manually trigger action'
on:
  workflow_dispatch:
    inputs:
      exit-code:
        type: number
        description: Override the EXIT_CODE
      kebab-bool:
        type: boolean
        description: Test kebab-case conditions
        default: true

jobs:
  lint-links-in-docs:
    name: 'Lint - Documentation Links (with issue reporting)'
    runs-on: ubuntu-22.04
    permissions:
      contents: read
      issues: write
    steps:
      - uses: actions/checkout@v3

      - id: link-reporter
        run: |
          echo '**report content here**' > "${{ runner.temp }}/report.md"

          # Only set when an explicit value is given to workflow input,
          EXIT_CODE=${{ github.event.inputs.exit-code }}
          # Otherwise fallback to 0:
          [[ -z ${EXIT_CODE} ]] && EXIT_CODE=0

          # Disabling this bool will set `EXIT_CODE=2`
          [[ ${{ github.event.inputs.kebab-bool }} == 'false' ]] && EXIT_CODE=2

          echo '::echo::on'
          echo "EXIT_CODE: ${EXIT_CODE}"

          # Test kebab-case output condition:
          echo "::set-output name=exit-code::${EXIT_CODE}"

      # Below manages issue creation/updating when failing links are found:

      - name: 'Check for an existing link-reporter issue'
        id: last-issue
        uses: micalevisk/last-issue-action@v1.2
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        with:
          state: open
          labels: ci-lint-links

      - name: 'Create a new issue, or update an existing one'
        if: ${{ steps.link-reporter.outputs.exit-code == 2 }}
        uses: peter-evans/create-issue-from-file@v4
        with:
          title: 'ci(lint): Docs have failing links'
          content-filepath: ${{ runner.temp }}/report.md
          issue-number: ${{ steps.last-issue.outputs.issue_number }}
          labels: ci-lint-links, bug
```

**Pull request closed without merging**
